backends/memcheck/ceed-memcheck-vector.c

```diff
@@ -174,44 +219,51 @@ static int CeedVectorRestoreArray_Memcheck(CeedVector vec) {
   CeedSize             length;
   CeedVector_Memcheck *impl;
 
+  CeedCallBackend(CeedVectorGetCeed(vec, &ceed));
   CeedCallBackend(CeedVectorGetData(vec, &impl));
   CeedCallBackend(CeedVectorGetLength(vec, &length));
-  CeedCallBackend(CeedVectorGetCeed(vec, &ceed));
 
-  memcpy(impl->array, impl->array_writable_copy, length * sizeof(CeedScalar));
-  for (CeedSize i = 0; i < length; i++) impl->array_writable_copy[i] = NAN;
-  CeedCallBackend(CeedFree(&impl->array_writable_copy));
+  // Check for unset entries after write-only access
   if (impl->is_write_only_access) {
     for (CeedSize i = 0; i < length; i++) {
-      if (isnan(impl->array[i]))
+      if (isnan(impl->array_writable_copy[i]))
         CeedDebug256(ceed, CEED_DEBUG_COLOR_WARNING, "WARNING: Vec entry %" CeedSize_FMT " is NaN after restoring write-only access", i);
     }
     impl->is_write_only_access = false;
   }
-  if (impl->array_borrowed) {
-    memcpy(impl->array_borrowed, impl->array, length * sizeof(CeedScalar));
-  }
-  if (impl->array_owned) {
-    memcpy(impl->array_owned, impl->array, length * sizeof(CeedScalar));
-  }
+
+  // Copy back to internal buffer and sync
+  memcpy(impl->array_allocated, impl->array_writable_copy, length * sizeof(CeedScalar));
+  CeedCallBackend(CeedVectorSyncArray_Memcheck(vec, CEED_MEM_HOST));
```

Comment:
It would be nice to drop this sync to help better catch memory access assumption violations, though that would require adding a bunch of empty calls to "SyncArray" or "TakeArray" in the CPU operator. But it would expand our number of potential bugs we could get ahead of for users.

Reply:
That would actually be a rather small perf impact if any, as the highest performance backends are children of Opt, and this change would be in Ref and Blocked. :thinking:

Reply:
Not for this PR, but to keep in mind